lib/Pipes/ModelInvalidationRule.cpp

```diff
@@ -0,0 +1,33 @@
+/// \file ModelInvalidationRule.cpp
+/// \brief Implementation of model invalidation rules
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "revng/Pipes/ModelInvalidationRule.h"
+
+using namespace revng::pipes;
+using namespace pipeline;
+
+llvm::ManagedStatic<std::vector<ModelInvalidationRule *>>
+  ModelInvalidationRule::Rules;
+
+using InvalidationMap = Runner::InvalidationMap;
+
+using EventType = ModelInvalidationEvent;
+void ModelInvalidationRule::allInvalidations(const Runner &Runner,
+                                             InvalidationMap &Map,
+                                             const EventType &Event) {
+  for (const auto &Step : Runner) {
+    auto &StepInvalidations = Map[Step.getName()];
+    for (const auto &Cotainer : Step.containers()) {
+      if (not Cotainer.second)
+        continue;
+
+      auto &ContainerInvalidations = StepInvalidations[Cotainer.first()];
+      for (const auto *Rule : *Rules)
+        Rule->getInvalidations(*Cotainer.second, ContainerInvalidations, Event);
+    }
+  }
```

Comment:
I've already said this in another comment but please consider shifting this responsibility to `Runner`, which already knows about invalidations. It might make the dependency graph between components cleaner.